### PR TITLE
vsphere: use packer to create the podvm template

### PR DESCRIPTION
**Pull request by whitebrandy**
v2: 
Added a few changes as pointed out in the reviews
 - Added the option to configure disk size by the user 
 - Cleaned up clutter in the Makefile and moved information gathering to a script
 - Renamed data/user-data.pkrtpl.hcl to data/user-data.pkr.hcl
 - Removed settings.auto.pkrvars.hcl, this will be created when building the template by the above script
 - Added *auto*.hcl files to .gitignore
 - Added a README
 - Fix for #339 (included in a previous push)

Mostly, derived from its libvirt counterpart, this script creates an esx guest from a standard ubuntu iso and then converts it into a template.

Standard settings are in settings.auto.pkrvars.hcl. vsphere config is expected in vsphere.auto.pkrvals.hcl which will be created by the Makefile if not present; user-data.pkrtpl.hcl contains the autoinstall template.

The automated input of characters at boot to start autoinstall is kind of flaky, the currently working sequence is defined in boot_command of the main script.

Fixes: #337 
Signed-off-by: Bandan Das <bsd@redhat.com>

### CODE REVIEW

**Comment by whitebrandy** [comment]
@caavery Please take a look and review when you get a chance.

**Comment by caavery** [comment]
Will do. I'm in the process of trying it out.

File: vsphere/image/files/etc/systemd/system/kata-agent.service [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Kata Agent
+After=network.target
+
+[Service]
+ExecStart=/usr/local/bin/kata-agent
```
Comment (fidencio) [inline vsphere/image/files/etc/systemd/system/kata-agent.service:6]: Something similar to https://github.com/confidential-containers/cloud-api-adaptor/pull/339 wil have to be done here, as in, passing the agent configuration and also creating the configuration file.

Comment (whitebrandy) [reply vsphere/image/files/etc/systemd/system/kata-agent.service]: Thanks for the review and the useful info. I updated kata-agent.service to reflect this change based on changes made in #339

**Comment by caavery** [comment]
Tested on my vsphere rig.

lgtm

File: vsphere/image/data/user-data.pkrtpl.hcl [hunk 2]
```diff
@@ -0,0 +1,198 @@
+#cloud-config
```
Comment (bpradipt) [inline vsphere/image/data/user-data.pkrtpl.hcl:1]: nit: can we keep the naming consistent with other packer files used in the project: `<name>.pkr.hcl` ? 
Ref: https://developer.hashicorp.com/packer/docs/templates/hcl_templates#code-organization)

This file could be just `user-data.pkr.hcl`

File: vsphere/image/settings.auto.pkrvars.hcl [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
+    DESCRIPTION:
```
Comment (bpradipt) [inline vsphere/image/settings.auto.pkrvars.hcl:2]: can some of the values be moved to `variables.pkr.hcl` as default values and this file be removed ?

File: vsphere/image/settings.auto.pkrvars.hcl [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+vm_guest_os_type = "ubuntu64Guest"
+
+
+// Virtual Machine Hardware Settings
+vm_firmware              = "efi"
+vm_cpu_count             = 2
+vm_mem_size              = 2048
+vm_disk_size             = 40960
+vm_disk_thin_provisioned = true
+vm_interface_name         = "vmxnet3"
```
Comment (bpradipt) [inline vsphere/image/settings.auto.pkrvars.hcl:19]: is the interface_name constant or depends on env ?

Comment (whitebrandy) [reply vsphere/image/settings.auto.pkrvars.hcl:19]: Sorry, this misnamed variable actually represents the virtualized network card. vmxnet3 is the most recent paravirtualized network card and is default in my setup but I am not sure if it really is! I believe it's best to specify to make sure we are consistent. Maybe, I can rename this variable to vm_network_card ?

File: vsphere/image/settings.auto.pkrvars.hcl (hunk 3 above)
Comment (whitebrandy) [reply vsphere/image/settings.auto.pkrvars.hcl:2]: The way I envisioned the design for this is that the *auto*.hcl files contain the user facing name=value settings that are most likely to change for a custom template build. Ideally, I would think that users should not have to alter the variables file to edit default values which has a more "structured" syntax and IMO is easier to break. The examples I followed from https://github.com/vmware-samples/packer-examples-for-vsphere also seem to follow this approach. Do you see any disadvantages to this ?

File: vsphere/image/data/user-data.pkrtpl.hcl (hunk 2 above)
Comment (whitebrandy) [reply vsphere/image/data/user-data.pkrtpl.hcl:1]: Will do, thanks for the review.

File: vsphere/image/settings.auto.pkrvars.hcl (hunk 3 above)
Comment (bpradipt) [reply vsphere/image/settings.auto.pkrvars.hcl:2]: Fair point. I was going by the pattern where any custom values will be passed as env variables and handled in the Makefile target. Example - https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/aws/image/Makefile#L73
If having the *.auto.pkrvars.hcl files then we need to decide whether to track it or add it to `.gitignore`. 

Another aspect to consider is the number of input params required. For a large number of params `*.auto.pkrvars.hcl` seems more convenient. Hence I was thinking if some of the defaults are common then those can be added to `variables.pkr.hcl` 
Anyways, I'll leave the decision to you. Both approaches are fine :-)

File: vsphere/image/settings.auto.pkrvars.hcl (hunk 4 above)
Comment (bpradipt) [reply vsphere/image/settings.auto.pkrvars.hcl:19]: You can add a brief explanation in the README

File: vsphere/image/settings.auto.pkrvars.hcl (hunk 3 above)
Comment (whitebrandy) [reply vsphere/image/settings.auto.pkrvars.hcl]: Thanks @bpradipt I pushed new changes and updated the cover letter to summarize them.

**Review by snir911 — APPROVED** [review]
I'm not a packer expert nor vsphere user, however from what i do understand, LGTM

**Review by caavery — APPROVED** [review]

**Pull request merged**